AdyenCardScanner/Sources/CardImageParser.swift

```diff
@@ -0,0 +1,233 @@
+//
+// Copyright (c) 2025 Adyen N.V.
+//
+// This file is open source and available under the MIT license. See the LICENSE file for more info.
+//
+
+import CoreImage
+import Foundation
+import Vision
+
+protocol CardImageParsing {
+    func parse(
+        image: CIImage,
+        completion: @escaping (CreditCard) -> Void
+    )
+}
+
+class CardImageParser: CardImageParsing {
+
+    private enum Constants {
+        static let expireDateRegex = "\\d{2}\\/\\d{2,4}"
+        static let topCandidates = 10
+
+        static let cardNumberConfidence: Float = 0.4
+        static let expireDateConfidence: Float = 0.4
+    }
+
+    // MARK: - Properties
+
+    private let dateFormatter: DateFormatter
+
+    private var cardNumber: String?
+    private var expireDate: Date?
+
+    // MARK: - Initializers
+
+    init() {
+        self.dateFormatter = DateFormatter()
+        self.dateFormatter.locale = Locale.current
+        self.dateFormatter.timeZone = TimeZone(secondsFromGMT: 0)
+    }
+
+    // MARK: - CardImageParsing
+
+    func parse(
+        image: CIImage,
+        completion: @escaping (CreditCard) -> Void
+    ) {
+        guard let transformedImage = transform(image: image) else { return }
+
+        let recognizeTextRequest = VNRecognizeTextRequest()
+        recognizeTextRequest.recognitionLevel = .accurate
+        recognizeTextRequest.usesLanguageCorrection = false
+
+        let imageRequestHandler = VNImageRequestHandler(ciImage: transformedImage, options: [:])
+        try? imageRequestHandler.perform([recognizeTextRequest])
+
+        guard let results = recognizeTextRequest.results, !results.isEmpty else {
+            return
+        }
+
+        let dispatchGroup = DispatchGroup()
+
+        dispatchGroup.enter()
+        DispatchQueue.global().async {
+            if self.cardNumber == nil {
+                self.cardNumber = self.extractCardNumber(from: results)
+            }
+            dispatchGroup.leave()
+        }
+
+        dispatchGroup.enter()
+        DispatchQueue.global().async {
+            if self.expireDate == nil {
+                self.expireDate = self.extractExpireDate(from: results)
+            }
+            dispatchGroup.leave()
+        }
+
+        dispatchGroup.notify(queue: .main) {
+            guard let cardNumber = self.cardNumber, let expireDate = self.expireDate else { return }
```

Comment:
The completion block is ignored on this case, leaving the caller unaware. Is this intended or otherwise we may need to call an empty/error state completion.